// @cycle/history/lib/cjs/captureClicks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var xstream_1 = require("xstream");
var CLICK_EVENT = typeof document !== 'undefined' && document.ontouchstart
    ? 'touchstart'
    : 'click';
function which(ev) {
    if (typeof window === 'undefined') {
        return false;
    }
    var e = ev || window.event;
    return e.which === null ? e.button : e.which;
}
function sameOrigin(href) {
    if (typeof window === 'undefined') {
        return false;
    }
    return href && href.indexOf(window.location.origin) === 0;
}

var baseHref = document.querySelector('base').getAttribute('href');

function makeClickListener(push) {
    return function clickListener(event) {
        if (which(event) !== 1) {
            return;
        }
        if (event.metaKey || event.ctrlKey || event.shiftKey) {
            return;
        }
        if (event.defaultPrevented) {
            return;
        }
        var element = event.target;
        while (element && element.nodeName !== 'A') {
            element = element.parentNode;
        }
        if (!element || element.nodeName !== 'A') {
            return;
        }
        if (element.hasAttribute('download') ||
            element.getAttribute('rel') === 'external') {
            return;
        }
        if (element.target) {
            return;
        }
        var link = element.getAttribute('href');
        if ((link && link.indexOf('mailto:') > -1) || link.charAt(0) === '#') {
            return;
        }
        if (!sameOrigin(element.href)) {
            return;
        }
        event.preventDefault();
        var pathname = element.pathname, search = element.search, _a = element.hash, hash = _a === void 0 ? '' : _a;

        // strip base href from the pathname. it gets added back by the history driver, which results
        // in having it twice without this fix.
        if (pathname.indexOf(baseHref) == 0) pathname = pathname.substr(baseHref.length-1);

        // temporary fix for https://github.com/cyclejs/cyclejs/issues/909, until a solution is merged into cycle
        push({ pathname, search, hash });
    };
}
function captureAnchorClicks(push) {
    var listener = makeClickListener(push);
    if (typeof window !== 'undefined') {
        document.addEventListener(CLICK_EVENT, listener, false);
    }
    return function () {
        return document.removeEventListener(CLICK_EVENT, listener);
    };
}
function captureClicks(historyDriver) {
    return function historyDriverWithClickCapture(sink$) {
        var cleanup;
        var internalSink$ = xstream_1.default.create({
            start: function () { },
            stop: function () { return typeof cleanup === 'function' && cleanup(); },
        });
        cleanup = captureAnchorClicks(function (loc) {
            internalSink$._n(Object.assign({ type: 'push' }, loc));
        });
        sink$._add(internalSink$);
        return historyDriver(internalSink$);
    };
}
exports.captureClicks = captureClicks;
